File: clients/python/examples/merkle_proof.py
"""Verify a compact Merkle inclusion proof (QDP-0010).

When a node publishes an anchor gossip message, it optionally attaches
a Merkle proof that binds the anchor transaction to a published
``Block.transactions_root``. Downstream peers verify the proof locally
— they do not need the full origin block.

This example builds a small 4-leaf tree, constructs a proof, and
verifies it. In production you'd pull the frames from
``AnchorGossipMessage.merkle_proof`` and the root from the
corresponding ``origin_block.transactions_root``.
"""

from __future__ import annotations

import hashlib

from quidnug import MerkleProofFrame, verify_inclusion_proof


def main() -> None:
    leaves = [hashlib.sha256(f"tx-{i}".encode()).digest() for i in range(4)]
    lvl0_pair0 = hashlib.sha256(leaves[0] + leaves[1]).digest()
    lvl0_pair1 = hashlib.sha256(leaves[2] + leaves[3]).digest()
    root = hashlib.sha256(lvl0_pair0 + lvl0_pair1).digest()

    # Build proof for tx at index 2. Its sibling at level 0 is leaves[3]
    # on the right side; at level 1 the sibling is the left subtree.
    frames = [
        MerkleProofFrame(hash=leaves[3].hex(), side="right"),
        MerkleProofFrame(hash=lvl0_pair0.hex(), side="left"),
    ]
    ok = verify_inclusion_proof(b"tx-2", frames, root.hex())
    print(f"Proof verifies: {ok}")

    # Sanity: a tampered leaf does not verify.
    tampered = verify_inclusion_proof(b"tx-forged", frames, root.hex())
    print(f"Tampered proof rejected: {not tampered}")


if __name__ == "__main__":
    main()

File: clients/rust/examples/merkle_proof.rs
//! QDP-0010 Merkle-proof verifier demo.

use quidnug::{verify_inclusion_proof, MerkleProofFrame};
use sha2::{Digest, Sha256};

fn sh(data: &[u8]) -> Vec<u8> {
    Sha256::digest(data).to_vec()
}

fn main() {
    let leaves: Vec<Vec<u8>> = (0..4).map(|i| sh(format!("tx-{i}").as_bytes())).collect();
    let pair0 = sh(&[&leaves[0][..], &leaves[1][..]].concat());
    let pair1 = sh(&[&leaves[2][..], &leaves[3][..]].concat());
    let root = sh(&[&pair0[..], &pair1[..]].concat());

    let frames = vec![
        MerkleProofFrame {
            hash: hex::encode(&leaves[3]),
            side: "right".into(),
        },
        MerkleProofFrame {
            hash: hex::encode(&pair0),
            side: "left".into(),
        },
    ];

    let ok = verify_inclusion_proof(b"tx-2", &frames, &hex::encode(&root)).unwrap();
    println!("valid proof: {}", ok);

    let tampered = verify_inclusion_proof(b"tx-forged", &frames, &hex::encode(&root)).unwrap();
    println!("tampered rejected: {}", !tampered);
}
